src/cljs/main/broadfcui/components/workspace_selector.cljs

```diff
@@ -2,41 +2,62 @@
+       (when (zero? selected-index)
+         ((@refs "new-workspace-form") :validate))))
+   :get-selected-workspace
+   (fn [{:keys [state refs]}]
+     (let [{:keys [workspaces selected-index]} @state]
+       (if (zero? selected-index)
+         {:new-workspace ((@refs "new-workspace-form") :get-field-values)}
+         {:existing-workspace (nth workspaces (dec selected-index))})))
+   :get-default-props
    (fn []
      {:filter identity
       :sort-by (comp (partial mapv string/lower-case)
                      (juxt :namespace :name)
                      :workspace)})
+   :get-initial-state
+   (fn []
+     {:selected-index 0})
    :render
    (fn [{:keys [props state locals]}]
-     (let [{:keys [workspaces]} @state
-           {:keys [on-select style]} props]
-       (assert on-select ":on-select is required for WorkspaceSelector")
+     (let [{:keys [workspaces selected-index]} @state
+           {:keys [style]} props]
        (if-not workspaces
          (spinner "Loading workspaces...")
-         (style/create-select
-           {:defaultValue ""
-            :ref (fn [elem]
-                   ;; doing this manually because the ref handler thingy calls :did-mount overzealously
-                   (when (and elem (not (:select @locals)))
-                     (swap! locals assoc :select elem)
-                     (.on (.select2 (js/$ elem)) "select2:select"
-                          (fn [event]
-                            (let [selected (nth workspaces (js/parseInt (.-value (.-target event))))]
-                              (on-select selected))))
-                     (on-select (first workspaces))))
-            :style (merge {:width 500} style)}
-           (map (fn [ws] (string/join "/" (replace (:workspace ws) [:namespace :name])))
-                workspaces)))))
+         [:div {}
+          (style/create-select
+            {:defaultValue ""
+             :ref (fn [elem]
+                    ;; doing this manually because the ref handler thingy calls :did-mount overzealously
+                    (when (and elem (not (:select @locals)))
+                      (swap! locals assoc :select elem)
+                      (.on (.select2 (js/$ elem)) "select2:select"
+                           (fn [event]
+                             (swap! state assoc :selected-index (-> event .-target .-value int))))))
+             :style (merge {:width 500} style)}
+            (->> workspaces
+                 (map (comp common/workspace-id->string :workspace))
+                 (cons "Create new workspace...")))
+          ;; Doing this via display: none to maintain state when the component is hidden
+          [:div {:style {:display (when-not (zero? selected-index) "none")
+                         :border style/standard-line
+                         :padding "0.5rem" :paddingBottom 0
+                         :margin "-0.5rem" :marginTop "0.5rem"}}
+           [ws-create/WorkspaceCreationForm {:ref "new-workspace-form"}]]])))
```

Comment:
Since string refs are being deprecated, do this with locals

Reply:
Then we really need to replace all of them across the application.

Reply:
Eventually, yes. For the time being, I'm only worrying about it with newly created refs.